### PR TITLE
Replace `getRepoURL` and `getRepoPath` with `getRepositoryInfo`

### PR DESCRIPTION
**Pull request by fregante**
Initially added in Refined GitHub, I think we can bring this over, unify it and deprecate any other function.

The 2 functions aren't mapped exactly to `getRepositoryInfo`. Here are the differences:

- `getRepoURL` did not actually check `isRepo`, so it would return `'settings'` if you were on `https://github.com/settings`
- `getRepoPath` did not use the canonical URL, however this probably isn't much of an issue generally

I need to ensure that this doesn't impact the current usage in RGH

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -506,8 +507,10 @@ const getRepoPath = (url: URL | HTMLAnchorElement | Location = location): string
 	return undefined;
 };
 
-/** Get the 'user/repo' part from an URL. Tries using the canonical URL to avoid capitalization errors in the `location` URL */
-const getRepoURL = (url?: URL | HTMLAnchorElement | Location): string => {
+/** Get the 'user/repo' part from an URL. Tries using the canonical URL to avoid capitalization errors in the `location` URL
+@deprecated Use getRepositoryInfo().url
+*/
+const getRepoURL = (url?: URL | Location): string => {
```
Comment (fregante) [inline index.ts:513]: Unlike `getRepoPath`, `getRepoURL` did not actually check `isRepoURL`, so I need to ensure that `getRepoURL` wasn't used improperly in RGH

**Review by fregante — COMMENTED** [review]
cc @yakov116 for any comments

I'll replace the usage of the deprecated functions both here and in RG before merging

File: index.ts [hunk 2]
```diff
@@ -516,9 +519,40 @@ const getRepoURL = (url?: URL | HTMLAnchorElement | Location): string => {
...
+		const canonical = document.querySelector<HTMLMetaElement>('[property="og:url"]'); // `rel=canonical` doesn't appear on every page
+		url = canonical ? canonical.content : location;
+	}
+
+	if (typeof url === 'string') {
+		url = new URL(url, location.origin);
+	}
+
+	if (!isRepo(url)) {
+		return {};
```
Comment (fregante) [inline index.ts:540]: I decided to return `{}` instead of `undefined` to make it easier to check without having to use `?.`

```js
// Before:
if (getRepoPath() === undefined) 

// Now:
if (getRepositoryInfo().path === undefined)

// If it returned `undefined`:
if (getRepositoryInfo()?.path === undefined)
```

**Comment by yakov116** [comment]
I debated this when I did the lint. One thing that held me back was that we sometimes need the repo url in lower case. So I backed off changing it all and just kept `getRepoURL` since I was not sure what needed it.

**Comment by fregante** [comment]
Good point. I added a couple of notes to my first post as well.

File: index.ts (hunk 2 above)
Comment (fregante) [reply index.ts]: Changed my mind. Returning `Partial<RepositoryInfo>` is actually wrong because the type is actually `RepositoryInfo | {}`.

The difference is that with the first one, even if you typecheck `if !repository.name, return`, `.path` will still be `string | undefined`, even if we know for sure it's a string.

Unfortunately `RepositoryInfo | {}` is awkward to use because `repository.name` will cause a type error: `name doesn't exist on {}`. 

In short, since we have `?.` in the language, let's make use of it for ease of use, even if it generates slightly longer code.

**Comment by fregante** [comment]
Pre-published as `4.3.0-0`

**Comment by fregante** [comment]
> we sometimes need the repo url in lower case

Do we? I think this was solved by our use of `og:url` in most cases. I'll try to replace our `getRepoURL` usage everywhere and we can review them in https://github.com/sindresorhus/refined-github/pull/3695

The only annoyance is that `getRepoURL()` becomes `getRepositoryInfo()!.url` and that it's a little slower because it runs more checks.

**Comment by yakov116** [comment]
> Do we?

https://github.com/sindresorhus/refined-github/blob/5b6b9c033e26da85ce9dc164801411da4d50b446/source/features/forked-to.tsx#L81

If I am not mistaken it was added because of that line. Will look for the PR to confirm.

**Comment by fregante** [comment]
Yeah, I mean we just need to make sure that all the comparisons are made on the same normalized content. In some cases we could still use `getRepositoryInfo()!.url.toLowerCase()` (😬), but not all.

**Comment by fregante** [comment]
Last 2 points:

- rename `url` to `nameWithOwner` to match GitHub’s wording and that it's not an actual `url`
- in most cases, this is "slower" than it needs to be since it introduces the normalization and `isRepo`… but I probably shouldn't worry about this
